### PR TITLE
Add support for multiple TeamTypes with limits/billing set per type

### PR DESCRIPTION
**Pull request by knolleary**
## Description

Part of #2358 

This expands the TeamTypes model to support the ability to have different TeamTypes with different limits/billing applied.

It is a significant piece of work as it required a lot of changes to how billing is structured given the configuration now has to come based on the TeamType.

I have structured the commits to try to make a bit more reviewable.

## DB Changes

 - `TeamType` - renames 'enabled' to 'active' to be consistent with 'ProjectTypes'
 - `TeamType` - adds 'order' column so the types can be listed in the desired order in the UI

The `Team` model has had a lot of instance functions added to it to access information about the team. This is a much cleaner approach. When running EE with billing, some of those functions are overloaded to add billing-specific checks and logic. This has removed the need for a lot of billing checks in the base CE code.

## API Changes

 - `/api/v1/team-types/***` - apis added to create/update/list/delete team types

## Config changes

Currently all team pricing config is provided in the yml file. With this PR, that info can now be set by admin in the UI.

When creating a TeamType, the admin has to provide a lot of information:

 - Stripe Price/Product for the TeamType itself
 - For each InstanceType on the platform - whether it is available to that Team, any limit on how many, a 'free allowance', and stripe price/product info
 - Device price/product
 - User limits (if any)
 - Trial mode options
 - Feature flags - although these are not currently honoured.

When the platform starts, it checks for the existing default 'starter' TeamType and adds if it necessary. If billing is enabled, it also adds the new Starter and Team (previously called Premium...) types - but leaves then inactive so the platform admin can choose when to enabled them (once they have finished configuring them).

#### Trial mode

Previously, trial mode was configured in admin settings - setting duration and what InstanceType is available. That configuration has moved to the individual TeamTypes configuration.

In Admin settings, if the 'create personal team' option is enabled, there is now a choice of what TeamType that personal team should be. It will default to the existing 'starter' team until configured otherwise.

---

What is missing before we can fully launch the new tiers in FF Cloud:

 - Upgrade path. There is currently no way for a team to change its type.
 - Feature limits - the UI lets the admin tick boxes to say what features should be enabled for a given type, but currently there are *no* restrictions applied
 - UX work around team creation/upgrade/billing. We have designs, but they are not yet implemented (and couldn't be until this base work was done).

### CODE REVIEW

**Comment by knolleary** [comment]
Postgres tests should be fixed with the most recent commit. The tests were doing a `TeamType.findOne()` to get the default team type - which is fine when there is only one. However, if billing is enabled, there will now be 3 team types - old starter, new Starter and new Premium. So the tests needed to be updated to ensure they got the old starter type.

File: forge/containers/wrapper.js [hunk 1]
```diff
@@ -241,9 +190,10 @@ module.exports = {
         }
     },
     restartFlows: async (project, options) => {
-        if (this._isBillingEnabled()) {
-            await this._subscriptionHandler.requireTrialProjectOrActiveSubscription(project)
-        }
+        // This will perform all checks needed to ensure this instance
+        // can be started - throws err if not
+        await project.Team.checkInstanceStartAllowed(project)
```
Comment (Pezmc) [inline forge/containers/wrapper.js:195]: That this needs a comment to explain how the method works each time feels like a code smell to me.

How about naming it `ensureInstanceStartAllowed`, or `throwIfInstanceStartNotAllowed`?

**Comment by knolleary** [comment]
I have just rebased to `main` and force pushed. This ensures the migration files are all present and correct in this branch following the merge of the pipeline work.

**Review by Pezmc — CHANGES_REQUESTED** [review]
Comments thus far, am about half way through a re-review. Just eating lunch!

File: forge/db/models/Team.js [hunk 2]
```diff
@@ -248,21 +248,112 @@ module.exports = {
...
                     return await M.Device.count({ where: { TeamId: this.id } })
+                },
+                /**
+                 * Many functions require this.TeamType to exist and be fully populated.
+                 * Depending on the route taken, it is possible this property has not
+                 * been fully loaded. This does the work to ensure it is there if needed.
+                 */
+                ensureTeamTypeExists: async function () {
+                    if (!this.TeamTypeId) {
+                        await this.reload({ include: [{ model: M.TeamType }] })
```
[the remark this answers is not in the document]
Comment (knolleary) [reply forge/db/models/Team.js:289]: There were multiple places where we load a model and include Team, but limit it to the minimum of team attributes that are needed in that context. That would mean it didn't have `TeamTypeId` because it wasn't one of the selected attributes.

The reload ensures the full Team object is retrieved and the TeamType can be included.

I did go through a lot of the relevant models to ensure they included `TypeTypeId`, but wasn't 100% confident there wasn't still a `reload` lurking somewhere that would get missed. This is a safety net just in case.

File: forge/db/models/Team.js [hunk 3]
```diff
@@ -248,21 +248,112 @@ module.exports = {
...
                 },
                 deviceCount: async function () {
                     return await M.Device.count({ where: { TeamId: this.id } })
+                },
+                /**
+                 * Many functions require this.TeamType to exist and be fully populated.
+                 * Depending on the route taken, it is possible this property has not
+                 * been fully loaded. This does the work to ensure it is there if needed.
+                 */
+                ensureTeamTypeExists: async function () {
```
[the remark this answers is not in the document]
Comment (knolleary) [reply forge/db/models/Team.js:287]: See comment below on why handling `TeamTypeId` being undefined is needed - if not ideal.

File: forge/db/views/TeamType.js [hunk 4]
```diff
@@ -19,36 +21,66 @@ module.exports = function (app) {
...
+        const result = {
             id: teamType.hashid,
             name: teamType.name,
+            active: teamType.active,
+            order: teamType.order,
             description: teamType.description,
             properties
         }
+        if (includeAdminOnlyProps && teamType.get) {
+            // For some API calls, the teamType passed here is a raw Object, not
```
[the remark this answers is not in the document]
Comment (knolleary) [reply forge/db/views/TeamType.js:78]: I agree this isn't ideal. I can't remember the specific code path that lead to needing to do this, but it was sufficiently non-trivial to resolve quickly that I worked around it.

File: forge/routes/api/teamType.js [hunk 5]
```diff
@@ -34,8 +31,251 @@ module.exports = async function (app) {
...
+        if (request.query.filter === 'all') {
+            filter = {}
+        } else if (request.query.filter === 'active') {
+            // Default behaviour
+            filter = { active: true }
+        } else if (request.query.filter === 'inactive') {
+            filter = { active: false }
+        }
+        const teamTypes = await app.db.models.TeamType.getAll(paginationOptions, filter)
+        teamTypes.types = teamTypes.types.map(pt => app.db.views.TeamType.teamType(pt, request.session.User.admin))
```
[the remark this answers is not in the document]
Comment (knolleary) [reply forge/routes/api/teamType.js:44]: A remnant of when I copy and pasted the boilerplate of a CRUD api from the **P**roject**T**ypes api.

File: forge/routes/api/device.js [hunk 6]
```diff
@@ -198,13 +198,8 @@ module.exports = async function (app) {
...
 
-        await team.reload({
-            include: [
-                { model: app.db.models.TeamType }
-            ]
-        })
-        const teamDeviceLimit = team.TeamType.getProperty('deviceLimit')
-        if (typeof teamDeviceLimit === 'number') {
+        const teamDeviceLimit = await team.getDeviceLimit()
+        if (teamDeviceLimit > -1) {
```
Comment (Pezmc) [inline forge/routes/api/device.js:202]: Personal preference `>= 0` (though is `0` valid?)

File: forge/routes/api/projectActions.js [hunk 7]
```diff
@@ -35,11 +35,13 @@ module.exports = async function (app) {
     }, async (request, reply) => {
         try {
             if (request.project.state === 'suspended') {
-                if (app.license.active() && app.billing) {
-                    if (!await app.billing.isProjectStartAllowed(request.project.Team, request.project)) {
-                        reply.code(402).send({ code: 'billing_required', error: 'Team billing not configured' })
-                        return
-                    }
+                try {
+                    // This will perform all checks needed to ensure this instance
```
Comment (Pezmc) [inline forge/routes/api/projectActions.js:39]: Re-reading these comments explaining what the function does is definitely a code smell, a clearer name should make them redundant!

File: frontend/src/pages/admin/Settings/General.vue [hunk 8]
```diff
@@ -245,6 +231,15 @@ export default {
...
         })
+        const teamTypes = await teamTypesApi.getTeamTypes()
+        this.teamTypes = teamTypes.types.map(tt => {
+            return {
+                order: tt.order,
+                value: tt.id,
+                label: tt.name
+            }
+        })
+        this.teamTypes.sort((A, B) => { return A.order - B.order })
```
Comment (Pezmc) [inline frontend/src/pages/admin/Settings/General.vue:242]: Usually this would go in a `computed` function, but we're not mutating `teamTypes` at any point, so this is ok for now.

**Comment by Pezmc** [comment]
I hit a few edge cases with this while testing https://github.com/flowforge/flowforge/pull/2553, all centered around adding a new instanceType _after_ the team type has been created. 

Fixes and test coverage are in [49efe7e](https://github.com/flowforge/flowforge/pull/2553/commits/49efe7ebe3138cb8098bce690132ce3ac552e3b0) and [40f6507](https://github.com/flowforge/flowforge/pull/2553/commits/40f650731f2a604a0dd3db991a727aca8c3cc9bd), but we should take extra care to verify editing team types before and after adding new instanceTypes on staging.

/CC: @MarianRaphael
